File: frontend_rebuild/types/engagement.ts
import type {
  Listing,
  PublicUser,
} from "@/types/listing";

import type {
  UserRole,
} from "@/types/auth";


export type InquiryType =
  | "general"
  | "inspection";

export type InquiryStatus =
  | "open"
  | "responded"
  | "closed";

export type ConversationListingType =
  | "rental"
  | "buy_property"
  | "marketplace";


/* =========================================================
   LEGACY INQUIRY
========================================================= */

export interface Inquiry {
  id: string;
  listing: Listing;
  sender: PublicUser;
  recipient: PublicUser;
  inquiry_type: InquiryType;
  message: string;
  status: InquiryStatus;
  responded_at: string | null;
  closed_at: string | null;
  created_at: string;
  updated_at: string;
}


/* =========================================================
   MESSAGES
========================================================= */

export interface InquiryParticipant {
  id: string;
  full_name: string;
  role: UserRole;
  profile_image_url: string | null;
}


export interface InquiryListingSummary {
  id: string;
  listing_type: ConversationListingType;
  title: string;
  slug: string;
}


export interface InquiryInboxItem {
  id: string;

  listing: InquiryListingSummary;
  other_user: InquiryParticipant;

  inquiry_type: InquiryType;
  status: InquiryStatus;

  last_message_preview: string;
  last_message_at: string;
  last_message_sender_id: string;

  unread: boolean;

  created_at: string;
}


export interface InquiryMessage {
  id: string;
  inquiry_id: string;
  sender_id: string;
  message: string;
  created_at: string;
  updated_at: string;
}


export interface InquiryThread {
  id: string;

  listing: InquiryListingSummary;

  sender: InquiryParticipant;
  recipient: InquiryParticipant;

  inquiry_type: InquiryType;
  status: InquiryStatus;

  responded_at: string | null;
  closed_at: string | null;

  created_at: string;

  messages: InquiryMessage[];

  has_more: boolean;
  next_before: string | null;
}

File: backend_rebuild/app/schemas/engagement.py
from datetime import datetime
from uuid import UUID

from pydantic import BaseModel, Field, field_validator

from app.models.enums import (
    InquiryStatus,
    InquiryType,
    ListingType,
    ReportReason,
    ReportStatus,
    UserRole,
)
from app.schemas.common import ORMModel
from app.schemas.listing import ListingRead
from app.schemas.user import PublicUserRead


class SavedListingRead(ORMModel):
    listing: ListingRead
    created_at: datetime


# ============================================================
# INQUIRIES / MESSAGES
# ============================================================


class InquiryCreate(BaseModel):
    inquiry_type: InquiryType = InquiryType.GENERAL
    message: str = Field(min_length=1, max_length=3000)

    @field_validator("message")
    @classmethod
    def clean_message(cls, value: str) -> str:
        cleaned = value.strip()

        if len(cleaned) < 10:
            raise ValueError(
                "Inquiry message must be at least 10 characters."
            )

        return cleaned


class InquiryStatusUpdate(BaseModel):
    status: InquiryStatus


class InquiryMessageCreate(BaseModel):
    message: str = Field(
        min_length=1,
        max_length=3000,
    )

    @field_validator("message")
    @classmethod
    def clean_message(cls, value: str) -> str:
        cleaned = value.strip()

        if not cleaned:
            raise ValueError(
                "Message cannot be empty."
            )

        return cleaned


class InquiryParticipantRead(BaseModel):
    id: UUID
    full_name: str
    role: UserRole
    profile_image_url: str | None


class InquiryListingSummaryRead(BaseModel):
    id: UUID
    listing_type: ListingType
    title: str
    slug: str


class InquiryInboxItemRead(BaseModel):
    id: UUID

    listing: InquiryListingSummaryRead
    other_user: InquiryParticipantRead

    inquiry_type: InquiryType
    status: InquiryStatus

    last_message_preview: str
    last_message_at: datetime
    last_message_sender_id: UUID

    unread: bool

    created_at: datetime


class InquiryMessageRead(BaseModel):
    id: UUID
    inquiry_id: UUID
    sender_id: UUID

    message: str

    created_at: datetime
    updated_at: datetime


class InquiryThreadRead(BaseModel):
    id: UUID

    listing: InquiryListingSummaryRead

    sender: InquiryParticipantRead
    recipient: InquiryParticipantRead

    inquiry_type: InquiryType
    status: InquiryStatus

    responded_at: datetime | None
    closed_at: datetime | None

    created_at: datetime

    messages: list[InquiryMessageRead]

    has_more: bool
    next_before: datetime | None


# ============================================================
# LEGACY / BACKWARDS-COMPATIBLE INQUIRY RESPONSE
# ============================================================


class InquiryRead(ORMModel):
    id: UUID
    listing: ListingRead
    sender: PublicUserRead
    recipient: PublicUserRead
    inquiry_type: InquiryType
    message: str
    status: InquiryStatus
    responded_at: datetime | None
    closed_at: datetime | None
    created_at: datetime
    updated_at: datetime


# ============================================================
# REPORTS
# ============================================================


class ListingReportCreate(BaseModel):
    reason: ReportReason
    details: str | None = Field(
        default=None,
        max_length=3000,
    )


class ListingReportResolve(BaseModel):
    status: ReportStatus
    resolution_note: str = Field(
        min_length=5,
        max_length=3000,
    )


class ListingReportRead(ORMModel):
    id: UUID
    listing: ListingRead
    reporter: PublicUserRead
    reason: ReportReason
    details: str | None
    status: ReportStatus
    resolution_note: str | None
    reviewed_by: PublicUserRead | None
    reviewed_at: datetime | None
    created_at: datetime
    updated_at: datetime